### ISSUE TITLE
Designing the software architecture for Project Cognoma

### ISSUE THREAD

**Issue by dhimmel** [cognoma/cognoma#5]
In #4 we've discussed some technical ideas and software tools. Here let's discuss the architecture of the 
Project Cognoma software stack. We came up with the following diagram: 

![cognoma](https://cloud.githubusercontent.com/assets/1117703/16665552/0ea33e00-4452-11e6-8740-ee48422c3a94.png)

You can modify the diagram by downloading [`cognoma.xml.txt`](https://github.com/cognoma/cognoma/files/352850/cognoma.xml.txt) and opening it at [draw.io](https://www.draw.io/).

Please ask questions and suggest modifications: the diagram is a rough draft and discussion will help us make improvements and fill in the details.

**Comment by awm33** [comment]
@dhimmel Why MongoDB for storing results?

**Comment by dhimmel** [comment]
We were thinking MongoDB rather than a relational database to give us more flexibility. Basically, we may want to add additional machine learning output fields in the future and thought it would be easier to build tolerance for missing fields into the results viewer than to have to migrate a database. Additionally, we may have some machine learning output that's simpler to store as JSON rather than as relational tables.

Does that make sense? I have little experience here.

**Comment by awm33** [comment]
@dhimmel Postgres can store JSON and has powerful query capability there. You can also do things like add schema to high level fields that are common and use JSON only for the differences. You would also be managing one less data storage technology.

**Comment by cgreene** [comment]
@awm33 That's pretty cool! Looks like django support in 1.9: https://docs.djangoproject.com/en/1.9/ref/contrib/postgres/fields/#jsonfield

**Comment by awm33** [comment]
Following up from @cgreene 's comment in https://github.com/cognoma/cognoma/issues/4

I've worked a lot on automated job/task management systems. Here's a proposal for a task service based on ones I've written and worked with. I'm calling it a service, but it could live within the main django app, or be separate service. I'm also willing to make this a contribution :)

![cognoma xml](https://cloud.githubusercontent.com/assets/3846080/16676623/390e241e-449a-11e6-8cc3-745ee7f6077a.png)

It would consist of:
- Postgres database tables, main one being the task table (or only one depending on implementation). A `unique` column can be used to prevent a job from be queued twice. The submitter would construct a string key that uniquely identifies a job (like "user-234-geneset-2387"). The `unique` column is combined with `task_type` into a unique partial index (postgres shinning here) where `status` is `queued` or `in_progress`. `run_at` and `schedule` columns can be used to create scheduled or recurling tasks.
- Something that workers would use to lock the task, this could be a message queue (like SQS, RabbitMQ, Cloud Pub/Sub, etc), Redis, or something like Zookeeper.

The communication between the app, workers, and task system could be direct (postgres+queue connections) with shared library code, or through a simple rest API. I tend to favor the rest approach, easier to implement in new code, just make some HTTP calls / long poll.

**Comment by awm33** [comment]
Since it looks like you want to expose the backend as a REST API, I would recommend writing the API as if you plan to make it public at some point. As in usable by other systems or UIs. Even if that never happens, it forces you to make better decisions. I used Flask once, but it looks like there is a nice django REST framework http://www.django-rest-framework.org/

I'm also wondering what you guys are thinking for frontend? Angular? React? Building your own styles from scratch or using something like Bootstrap?

**Comment by dcgoss** [comment]
Django Rest Framework is fantastic and I would strongly recommend it for a Django API.

**Comment by cgreene** [comment]
@awm33: This sounds great! I'd hope to make the API public along with the web app. We have some Angular experience within the lab, but I'd be comfortable with either (React, Angular) based on where we have the most strength based on contributors.

@dcgoss: That's the second item of high praise I've heard for django-rest-framework. Sounds like a great way to go!

**Comment by mydpy** [comment]
A few data questions: 
1. What is the expected volume for each data store? (reads/writes per time grain) 
2. How are we storing results from the ML engine? Do you imagine serialized JSON blobs, or something else? 
3. Is it out-of-scope to discuss ingesting Cancer data in this model? Are you getting data from sources in real-time or are we pre-processing existing data stores?

**Comment by cgreene** [comment]
1. Depends a bit on usage.
2. My inclination is JSON that the results viewer loads, but I am not strongly tied to that.
3. We will need to ingest data - it probably won't be in real time (the resources we are using are updated with only modest frequency).

Thanks!
Casey

**Comment by tantanchen** [comment]
My limited understanding of ML is that the algorithms take a while to run, but once a model is complete, it is very quick to use in predictions. So what languages will we use to train the models? and how will we store the models?

**Comment by dhimmel** [comment]
> My limited understanding of ML is that the algorithms take a while to run, but once a model is complete, it is very quick to use in predictions.

Yes, a simple but important insight -- training a model can be laborious, but applying the model for predictions is usually fast! The one caveat is that often computational time is not the bottleneck. Rather human time troubleshooting models, exploring data, and refining the setup takes the bulk of the time. Hopefully, we can find a setup that produces decent models for the domain we're modeling.

> So what languages will we use to train the models?

I was thinking we'd start with [`sklearn`](http://scikit-learn.org/stable/) in Python. The API is standardized across algorithms and relatively mature. However, we're aiming for a modular design, so if there is another language/package that offers an interesting algorithm, we should be able to include it.

> how will we store the models

We'll want to define several model attributes that pertain to all machine learning algorithms. For example, these include predictions on the training data and the corresponding performance (AUROC & AUPRC, Tjur for probabilities). However, each algorithm will have algorithm-specific info, such as feature coefficients for logistic regression. I was thinking we would [use JSON](#issuecomment-231186190) here, because it's versatile.

**Comment by zfanelle** [comment]
Hey guys, great ideas above. I have some ideas as well.

I am unfamiliar with the type of data we will be storing in the Mongodb database. Do we have a sample data model entry? I would recommend staying the in relational world if we have to (for cleaner, more enforceable data).

I would recommend Mongodb if we:
- Require a schema-less data model (Having a stricter relational model will help us keep our data cleaner)
- Require horizontal scalability and high I/O

If we want a flexible data model (we know that we will be adding columns all the time) If we are set on JSON, MongoDb is a good idea like suggested above. A good column store would be Cassandra. I can write ansible playbooks for automation and documentation for the databases. We could also get a hosted mongoDB for very cheap: https://mlab.com/ .I can also write ansible playbooks for other servers as well (like the app servers and load balancers).

As for the Postgres SQL, I think that is a good idea. I also recommend MySQL as well as it is more prevalent and just as performant and flexible. For either of these, we could use Amazon RDS to host these. If we don't want to be locked in a vendor, we could export our data and move somewhere else. I had a great experience with RDS for stability, performance, and disaster recovery. We could host our own in a VM as well.

**Comment by stevmoli** [comment]
Hey, could somebody please give me some clarification: am I correct in thinking that the Postgres Database is going to be used for storing a table of jobs/tasks queued by the users?  Thanks!

**Comment by awm33** [comment]
@stevmoli Yes :) Here is the repo containing those models https://github.com/cognoma/task-service

**Comment by stevmoli** [comment]
Awesome, thanks for the clarification!
